server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -322,7 +325,13 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
   {
     try {
       log.info("Loading segment %s", segment.getIdentifier());
-      if (loadSegment(segment, callback)) {
+      if (segmentsToDelete.contains(segment)) {
+        synchronized (lock) {
+          segmentsToDelete.remove(segment);
+        }
+      }
```

Comment:
The synchronize block is used to prevent race conditions happened when the scheduled delete runnable starts.
Specifically, when the runnable starts, 
and ```if (segmentsToDelete.contains(segment))``` returns true, historical will start dropping segment files. However, it's possible that right after the "if" check, addSegment() is called and actually loads the segment, which makes dropping segment unncessary.